include/kframework/lemmas/int-simplification.k

```diff
@@ -158,4 +161,48 @@ module INT-SIMPLIFICATION-COMMON
 
     rule A -Int B +Int C <=Int D => false requires D <Int A -Int B andBool 0 <=Int C [simplification]
 
+  // ###########################################################################
+  // up/Int
+  // ###########################################################################
+
+    rule [upInt-lt-true]:
+      ((X up/Int Y) *Int Y) <Int Z => true
+      requires X +Int Y <=Int Z andBool 0 <Int Y
+      [simplification]
+
+    rule [upInt-lt-false]:
+      ((X up/Int Y) *Int Y) <Int Z => false
+      requires Z <=Int X andBool 0 <Int Y
+      [simplification]
+
+    rule [upInt-refl-lt]:
+      ((X up/Int Y) *Int Y) <Int X => false
+      requires 0 <Int Y
+      [simplification]
+
+    rule [upInt-refl-leq]:
+      ((X up/Int Y) *Int Y) <=Int X => X modInt Y ==Int 0
+      requires 0 <Int Y
+      [simplification]
+
+    rule [upInt-refl-gt]:
+      X <Int ((X up/Int Y) *Int Y) => X modInt Y =/=Int 0
+      requires 0 <Int Y
+      [simplification]
+
+    rule [upInt-refl-geq]:
+      X <=Int ((X up/Int Y) *Int Y) => true
+      requires 0 <Int Y
+      [simplification]
```

Comment:
I assume that this is one of the lemmas that breaks z3 if it's made an `smt-lemma`, right? Because if we had this as an `smt-lemma` and also another that said `((X up/Int Y) *Int Y) <Int X + Y => true`, then we wouldn't need `upInt-lt-true` and `upInt-lt-false`

Reply:
Yes, I think that's correct, we wouldn't need those. 

Any lemma with `up/Int` will break Z3 if `smt-lemma`' is added. I think this is because Z3 has no idea what `up/Int` actually is and then when faced with constraints it just goes 'I don't know!'.

Reply:
I'm looking at the definition of `up/Int` in [`evm-types.md`](https://github.com/runtimeverification/evm-semantics/blob/master/include/kframework/evm-types.md) and it's defined with an `smtlib` attribute, which should give it uninterpreted function semantics. I don't get why this only happens to `up/Int` and not any of the other functions with the `smtlib` attribute. Does @ehildenb know anything about this?

Reply:
It doesn't only happen to `up/Int`, I can also break it by requiring `smt-lemma` on the facts that `&Int` cannot be greater than its operands...

Reply:
It might just be the limit of Z3 reasoning that we're hitting...

Reply:
We could try alternative Z3 theories. For example, add the `smt-lemma`, then get hte Z3 transcript, then at the top it specifies which Z3 theories it's reasoning in. We may be able to just change the theory support it has and suddenly get better results.

Reply:
The transcript I got doesn't have the theories, it starts with:
```
(set-option :print-success true )
; success

(set-option :timeout 40 )
; success

(declare-sort |HB_#String| 0 )
; success

(declare-sort |HB_SortAccessedAccountsCellOpt| 0 )
; success

(declare-sort |HB_SortAccessedStorageCellOpt| 0 )
; success

...
```
and then just continues with the declarations and assertions...

Reply:
Hmmmm, that probably means we're just using Z3's default theory, which is probably not very useful. I wonder if we could try some different Z3 theories on that transcript you have?

Here is an example of setting a different logic: https://microsoft.github.io/z3guide/docs/theories/Arithmetic/#difference-arithmetic, and here are a bunch of supported logics: https://stackoverflow.com/questions/23450572/which-logics-are-supported-by-z3

These logics should really be called "theories", but whatever.